src/Processors/Formats/Impl/NativeORCBlockInputFormat.cpp

```diff
@@ -1285,52 +1286,48 @@ static ColumnPtr readOffsetsFromORCListColumn(const BatchType * orc_column)
+readColumnWithBooleanData(const orc::ColumnVectorBatch * orc_column, const String & column_name)
 {
     const auto * orc_bool_column = dynamic_cast<const orc::LongVectorBatch *>(orc_column);
     auto internal_type = DataTypeFactory::instance().get("Bool");
     auto internal_column = internal_type->createColumn();
     auto & column_data = assert_cast<ColumnVector<UInt8> &>(*internal_column).getData();
-    column_data.reserve(orc_bool_column->numElements);
 
+    column_data.reserve(orc_bool_column->numElements);
     for (size_t i = 0; i < orc_bool_column->numElements; ++i)
-        column_data.push_back(static_cast<UInt8>(orc_bool_column->data[i]));
+    {
+        if (!orc_bool_column->hasNulls || orc_bool_column->notNull[i])
+            column_data.push_back(static_cast<UInt8>(orc_bool_column->data[i]));
+        else
+            column_data.push_back(0);
+    }
 
     return {std::move(internal_column), internal_type, column_name};
 }
 
-/// Inserts numeric data right into internal column data to reduce an overhead
-template <typename NumericType, typename BatchType, typename VectorType = ColumnVector<NumericType>>
-static ColumnWithTypeAndName
-readColumnWithNumericData(const orc::ColumnVectorBatch * orc_column, const orc::Type *, const String & column_name)
-{
-    auto internal_type = std::make_shared<DataTypeNumber<NumericType>>();
-    auto internal_column = internal_type->createColumn();
-    auto & column_data = static_cast<VectorType &>(*internal_column).getData();
-    column_data.reserve(orc_column->numElements);
-
-    const auto * orc_int_column = dynamic_cast<const BatchType *>(orc_column);
-    column_data.insert_assume_reserved(orc_int_column->data.data(), orc_int_column->data.data() + orc_int_column->numElements);
 
-    return {std::move(internal_column), internal_type, column_name};
-}
-
-template <typename NumericType, typename BatchType, typename VectorType = ColumnVector<NumericType>>
+template <typename NumericType, typename BatchType>
 static ColumnWithTypeAndName
-readColumnWithNumericDataCast(const orc::ColumnVectorBatch * orc_column, const orc::Type *, const String & column_name)
+readColumnWithNumericData(const orc::ColumnVectorBatch * orc_column, const String & column_name)
 {
     auto internal_type = std::make_shared<DataTypeNumber<NumericType>>();
     auto internal_column = internal_type->createColumn();
-    auto & column_data = static_cast<VectorType &>(*internal_column).getData();
-    column_data.reserve(orc_column->numElements);
-
+    auto & column_data = static_cast<ColumnVector<NumericType> &>(*internal_column).getData();
     const auto * orc_int_column = dynamic_cast<const BatchType *>(orc_column);
+
+    column_data.reserve(orc_int_column->numElements);
     for (size_t i = 0; i < orc_int_column->numElements; ++i)
-        column_data.push_back(static_cast<NumericType>(orc_int_column->data[i]));
+    {
+        if (!orc_int_column->hasNulls || orc_int_column->notNull[i])
```

Comment:
ditto

Reply:
The problem was that the ORC library didn't initialize the memory in the array with values when it was null. Then we memcpy this array into our column, which is fast, and as a column is Nullable, this uninitialized data is not visible. Until we expose it, e.g., when serializing in Native format and sending over network.

Reply:
I don't think this is a problem worth worrying about. When `nullmap[i]` is 1, The value of `data[i]` does not affect the semantics. CH don't have a strong guarantee that `data[i]` must be default value when `nullmap[i]` is 1 in `ColumnNullable`

Reply:
@taiyang-li Memory Sanitizer has its word: https://s3.amazonaws.com/clickhouse-test-reports/json.html?PR=85063&sha=9fd2d94c75a3f75c48fcff10f87fce1600041811&name_0=PR